lib/spack/docs/tutorial_buildsystems.rst

```diff
@@ -0,0 +1,351 @@
+fact, for most of the build systems, overriding methods is optional and packages
+can be kept as simple as what we have already. 
+
+Unfortunately, the defaults that bowtie sets up are not what we 
+need. If we look inside the Makefile, we see that CC and CXX point to our GNU
+compiler:
+
+.. code-block:: make
+    
+    CPP = g++ -w
+    CXX = $(CPP)
+    CC = gcc
+    LIBS = $(LDFLAGS) -lz
+    HEADERS = $(wildcard *.h)
+
+To fix this we need to use the edit method which is used to write to our custom
+Makefile:
+
+.. literalinclude:: tutorial/examples/Makefile/2.package.py
+   :language: python
+
+Here we used regex to make CC and CXX point at whatever Spack sets
+CC and CXX to. This allows us to build bowtie with whatever compiler we specify
+through Spack's specs. We can also specify our build targets and install targets
+usng the build() and install() methods.
+
+.. literalinclude:: tutorial/examples/Makefile/3.package.py
+   :language: python
+
+In our build stage we are able to provide an argument to make whether we
+decide to build with tbb support or not. For the install, phase we provide
+the prefix and then invoke make install. 
+
+Through this example, we are able to see that we can change the Makefile through
+our edit method, as well as invoke build, and install to specify our build
+and install targets, respectively.
+
+-------------------
+Autotools
+-------------------
+
+As mentioned earlier, Autotools consist of three parts:  autoconf, automake,
+and libtool. Your typical build incantation will consist of the following:
+
+.. code-block:: console
+
+    ./configure --prefix=/path/to/install_dir
+    make
+    make install
+
+Although it is possible to write these packages using the vanilla Package 
+class, Spack has an Autotools class that automatically sets up common 
+arguments to configure. In the Autotools class we can override four phases: 
+autoreconf, configure, build, and install. Autoreconf is only necessary if 
+there is no configure script present and one has to be generated using. 
+Most packagers will only need to override the helper method 
+configure_args since Spack provides sensible defaults for each phase. 
+
+Let's look at the mpileaks package.py file that we worked on earlier:
+
+
+.. code-block:: console
+    
+    $ spack edit mpileaks
+
+Notice that mpileaks is a Package class but uses the Autotools build system 
+(the configure, make, make install incantation gives us a hint). Although
+this package is acceptable lets make this into an AutotoolsPackage class and
+simplify it further.
+
+.. literalinclude:: tutorial/examples/Autotools/0.package.py
+
+We first inherit from the AutotoolsPackage class. 
+
+We can keep the install method but we can simplify this further by adding configure_args:
+
+.. literalinclude:: tutorial/examples/Autotools/1.package.py
+
+A couple of things to note here: We got rid of "prefix=" + prefix, and also
```

Comment:
We should probably show the AutotoolsPackage class here so that they can see that the prefix is added automatically.

And in general I think we need to show the Package subclasses so that they can see what the subclasses do for them. I think that's easier to show than explain.